File: src/面试题08_10.color-fill-lcci.100201/solution.ts
function floodFill (image: number[][], sr: number, sc: number, newColor: number): number[][] {
    dfs(image, sr, sc, image[sr][sc], newColor);
    return image;
}

function dfs (image:number[][], x:number, y:number, oldColor:number, newColor:number) {
    if (x < 0 || y < 0 || x === image.length || y === image[0].length || image[x][y] !== oldColor || image[x][y] === newColor) {
        return;
    }
    image[x][y] = newColor;
    dfs(image, x - 1, y, oldColor, newColor);
    dfs(image, x + 1, y, oldColor, newColor);
    dfs(image, x, y - 1, oldColor, newColor);
    dfs(image, x, y + 1, oldColor, newColor);
}

File: src/面试题08_10.color-fill-lcci.100201/solution.java
class Solution {
    int originColor;
    int R;
    int C;
    int[][] image;
    int newColor;
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        originColor = image[sr][sc];
        R = image.length;
        C = image[0].length;
        this.image = image;
        this.newColor = newColor;
        if(originColor == newColor){
            return image;
        }
        dfs(sr,sc);

        return image;
    }

    private void dfs(int r,int c){
        if(r == -1 || c == -1 || r == R || c == C || image[r][c] != originColor){
            return;
        }
        image[r][c] = newColor;
        
        dfs(r-1,c);
        dfs(r+1,c);
        dfs(r,c-1);
        dfs(r,c+1);
        
    }
}